PagoPAGetUser/handler.ts

```diff
@@ -0,0 +1,143 @@
+import {
+  IResponseErrorQuery,
+  ResponseErrorQuery
+} from "io-functions-commons/dist/src/utils/response";
+import {
+  IResponseErrorNotFound,
+  IResponseErrorValidation,
+  IResponseSuccessJson,
+  ResponseErrorNotFound,
+  ResponseErrorValidation,
+  ResponseSuccessJson
+} from "italia-ts-commons/lib/responses";
+import { EmailString } from "italia-ts-commons/lib/strings";
+import { PagoPAUser } from "../generated/definitions/PagoPAUser";
+import { User } from "../types/user";
+import { RequiredExpressUserMiddleware } from "../utils/middleware/required_express_user";
+import SessionStorage from "../utils/sessionStorage";
+
+type IHttpHandler = (
+  context: Context,
+  user: User
+) => Promise<IPagoPaGetUserHandlerResult>;
+
+type IPagoPaGetUserHandlerResult =
+  | IResponseSuccessJson<PagoPAUser>
+  | IResponseErrorNotFound
+  | IResponseErrorValidation
+  | IResponseErrorQuery;
+
+export function PagoPaGetUserHandler(
+  profileModel: ProfileModel,
+  sessionStorage: SessionStorage,
+  enableNoticeEmailCache: boolean
+): IHttpHandler {
+  return async (_CTX, user) => {
+    const getProfileAndSaveNoticeEmailCache = profileModel
+      .findLastVersionByModelId([user.fiscal_code])
+      .foldTaskEither<
+        IResponseErrorNotFound | IResponseErrorQuery,
+        Option<EmailString>
+      >(
+        failure =>
+          fromLeft(
+            ResponseErrorQuery("Error while retrieving the profile", failure)
+          ),
+        maybeProfile =>
+          maybeProfile.fold<
+            TaskEither<IResponseErrorNotFound, Option<EmailString>>
+          >(
+            fromLeft<IResponseErrorNotFound, Option<EmailString>>(
+              ResponseErrorNotFound(
+                "Profile not found",
+                "The profile you requested was not found in the system."
+              )
+            ),
+            profile => {
+              const maybeNoticeEmail: EmailString | undefined =
+                profile.email && profile.isEmailValidated
+                  ? profile.email
+                  : user.spid_email;
+              return taskEither.of(fromNullable(maybeNoticeEmail));
+            }
+          )
+      ) // Save the value into the redis cache
+      .chain(maybeNoticeEmail => {
+        if (isNone(maybeNoticeEmail)) {
+          return taskEither.of(maybeNoticeEmail as Option<EmailString>);
+        }
+        return tryCatch(
+          () =>
+            sessionStorage.setPagoPaNoticeEmail(user, maybeNoticeEmail.value),
+          () => new Error("Error caching the notify email value")
+        ).foldTaskEither(
+          _1 => taskEither.of(maybeNoticeEmail),
+          _1 => taskEither.of(maybeNoticeEmail)
+        );
+      });
+    return (enableNoticeEmailCache
+      ? tryCatch(
+          () => sessionStorage.getPagoPaNoticeEmail(user),
+          _ => new Error("Error reading the notify email cache")
+        )
+          .foldTaskEither(_ => fromLeft<Error, EmailString>(_), fromEither)
```

Comment:
I think this is enough:
```suggestion
        ).map(fromEither)
```

Reply:
This line is needed to handle the `Either` on the right and merge the two left scenarios on a single path.

Reply:
sorry, actually it was `chain`

```suggestion
        ).chain(fromEither)
```

this works, same result

Reply:
Nice, it works fine. Fixed in 15a4a73